scripts/copy-licences.py

```diff
@@ -0,0 +1,128 @@
+# (C) Copyright 2024- ECMWF.
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# In applying this licence, ECMWF does not waive the privileges and immunities
+# granted to it by virtue of its status as an intergovernmental organisation nor
+# does it submit to any jurisdiction.
+#
+
+import json
+import re
+import sys
+
+import requests
+from html2text import html2text
+
+
+def identity(x):
+    return x
+
+
+ENTRIES = {
+    "libeccodes": {
+        "home": "https://github.com/ecmwf/eccodes",
+        "copying": "https://raw.githubusercontent.com/ecmwf/eccodes/develop/LICENSE",
+    },
+    "libpng": {
+        "home": "https://github.com/glennrp/libpng",
+        "copying": "https://raw.githubusercontent.com/glennrp/libpng/libpng16/LICENSE",
+    },
+    "libaec": {
+        "home": "https://github.com/MathisRosenhauer/libaec",
+        "copying": "https://raw.githubusercontent.com/MathisRosenhauer/libaec/master/LICENSE.txt",
+    },
+    "libjasper": {
+        "home": "https://github.com/jasper-software/jasper",
+        "copying": "https://raw.githubusercontent.com/jasper-software/jasper/master/LICENSE.txt",
+    },
+    "libopenjp2": {
+        "home": "https://github.com/uclouvain/openjpeg",
+        "copying": "https://raw.githubusercontent.com/uclouvain/openjpeg/master/LICENSE",
+    },
+    "libopenjpeg": {
+        "home": "https://code.google.com/archive/p/openjpeg/",
+        "copying": "https://raw.githubusercontent.com/uclouvain/openjpeg/master/LICENSE",
+    },
+    "libjpeg": {
+        "home": "http://ijg.org",
+        "copying": "https://jpegclub.org/reference/libjpeg-license/",
+        "html": True,
+    },
+}
+
+PATTERNS = {
+    r"^libpng\d+$": "libpng",
+    r"^libproj(_\d+)+$": "libproj",
+}
+
+ALIASES = {
+    "libeccodes_memfs": "libeccodes",
+}
+
+if False:
+    for e in ENTRIES.values():
+        if isinstance(e, dict):
+            copying = e["copying"]
+            if copying.startswith("http"):
+                requests.head(copying).raise_for_status()
+
+libs = {}
+missing = []
+seen = set()
+
+for line in open(sys.argv[1], "r"):  # noqa: C901
+    lib = "-no-regex-"
+    lib = line.strip().split("/")[-1]
+    lib = lib.split("-")[0].split(".")[0]
+
+    if lib == "":
+        continue
+
+    if not lib.startswith("lib"):
+        lib = f"lib{lib}"
+
+    for k, v in PATTERNS.items():
+        if re.match(k, lib):
+            lib = v
+
+    lib = ALIASES.get(lib, lib)
+
+    if lib not in ENTRIES:
+        missing.append((lib, line))
+        continue
+
+    if lib in seen:
+        continue
+
+    seen.add(lib)
+
+    e = ENTRIES[lib]
+    if e is None:
+        continue
+
+    libs[lib] = dict(path=f"copying/{lib}.txt", home=e["home"])
+    copying = e["copying"]
+
+    filtering = identity
+    if e.get("html"):
+        filtering = html2text
+
+    with open(f"eccodes/copying/{lib}.txt", "w") as f:
```

Comment:
this eventually makes up the licenses end up in the package itself, rather than its distinfo, right? (I checked that it behaves like that in ecmwf-libs). I guess it would be more proper to put those third party licenses to distinfo too, perhaps as a `thirdy_party/LICENSE_BUNDLED.txt` or a collection of `LICENSE` files (I'm drawing my info eg on this discussion https://github.com/pytorch/pytorch/issues/81181 and the linked PEP, not that I'm a licensing expert :) )

or it may actually be even simpler -- we are compiling without the jpeg/jasper support for eccodes, right? So is there even any third party library used here? If not we should be fine with just the regular license file at the repo root, shouldnt we?

Reply:
Almost certainly you will need libaec to build ecCodes for mir-python, but you can very likely disable png and jpeg support. Thanks for the extra info on licences - I will have a look.

Reply:
Hi @tmi, possibly the reason that ecmwflibs does not put the copyright info into distinfo is that it contains a function to display the various copyright notices, so it has to be able to read them in. (https://github.com/ecmwf/ecmwflibs/blob/master/ecmwflibs/__init__.py#L182). It would be a bit harder to figure out the path if it was in distinfo, so I'm guessing this is why it was done this way. Although I have not included this function in ecCodes, I may add it in the future, so I think I'll keep these files where they are, at least for now.